// @timberio/ui/dist/components/Step/Step.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactEmotion = _interopRequireDefault(require("react-emotion"));

var _helpers = require("../../helpers");

var _Flex = _interopRequireDefault(require("../Box/Flex"));

var _Box = _interopRequireDefault(require("../Box"));

var _Icon = _interopRequireDefault(require("../Icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Divider =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "e78y5du0"
})("border-left:3px solid ", (0, _helpers.fromTheme)('colors.gray.2'), ";flex-grow:1;");

var Step = function Step(_ref) {
  var icon = _ref.icon,
      children = _ref.children,
      bg = _ref.bg,
      rest = _objectWithoutProperties(_ref, ["icon", "children", "bg"]);

  return _react.default.createElement(_Flex.default, rest, _react.default.createElement(_Flex.default, {
    alignItems: "center",
    flexDirection: "column"
  }, _react.default.createElement(_Flex.default, {
    borderRadius: "50%",
    bg: bg,
    boxShadow: 0
  }, _react.default.createElement(_Icon.default, {
    name: icon,
    color: "#FFF",
    size: 14,
    m: 10
  })), _react.default.createElement(Divider, null)), _react.default.createElement(_Box.default, {
    pt: 9,
    px: 3,
    flex: "1 1 auto"
  }, children));
};

Step.defaultProps = {
  bg: 'purple.3'
};
Step.propTypes = {
  icon: _propTypes.default.string,
  children: _propTypes.default.node,
  bg: _propTypes.default.string
};
var _default = Step;
exports.default = _default;
