// @timberio/ui/dist/components/MenuItem/MenuItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _cleanElement = _interopRequireDefault(require("clean-element"));

var _objectPath = require("object-path");

var _helpers = require("../../helpers");

var _Icon = _interopRequireDefault(require("../Icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var cleaned = (0, _cleanElement.default)('div');
var Wrapper = (0, _helpers.system)(cleaned)(function (props) {
  return {
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    color: (0, _objectPath.get)(props, 'theme.colors.gray.5'),
    fontSize: '.9rem',
    padding: '.75rem',
    ':hover': {
      background: (0, _objectPath.get)(props, 'theme.colors.gray.1'),
      color: (0, _helpers.intentColor)(props, {
        intent: 'primary',
        shade: 3
      })
    }
  };
});

var MenuItem = function MenuItem(_ref) {
  var icon = _ref.icon,
      children = _ref.children,
      rest = _objectWithoutProperties(_ref, ["icon", "children"]);

  return _react.default.createElement(Wrapper, rest, icon && _react.default.createElement(_Icon.default, {
    name: icon,
    size: 16,
    mr: 2,
    fallback: "box"
  }), children);
};

MenuItem.propTypes = {
  icon: _propTypes.default.string
};
cleaned.propTypes = MenuItem.propTypes;
MenuItem.displayName = 'MenuItem';
var _default = MenuItem;
exports.default = _default;
