// @timberio/ui/dist/components/Ranger/components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Handle = exports.Segment = exports.TickLabel = exports.Tick = exports.Track = void 0;

var _reactEmotion = _interopRequireDefault(require("react-emotion"));

var _helpers = require("../../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Track =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "etmia1v0"
})("display:inline-block;height:", function (props) {
  return props.height;
}, ";width:90%;background:", (0, _helpers.fromTheme)('colors.gray.2'), ";");
exports.Track = Track;
var Tick =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "etmia1v1"
})(":before{content:\"\";position:absolute;left:0;background:rgba(0,0,0,0.2);height:5px;width:2px;transform:translate(-50%,0.7rem);}");
exports.Tick = Tick;
var TickLabel =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "etmia1v2"
})("font-size:0.7rem;color:", (0, _helpers.fromTheme)('colors.gray.3'), ";top:100%;transform:translate(-50%,1.2rem);white-space:nowrap;");
exports.TickLabel = TickLabel;
var Segment =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "etmia1v3"
})("height:100%;");
exports.Segment = Segment;
var Handle =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "etmia1v4"
})("background:", (0, _helpers.fromTheme)('colors.purple.3'), ";display:flex;align-items:center;justify-content:center;width:", function (props) {
  return props.width;
}, ";height:", function (props) {
  return props.height;
}, ";border-radius:", function (props) {
  return props.radius;
}, ";font-size:.9rem;font-weight:bold;white-space:nowrap;color:white;cursor:pointer;transform:", function (props) {
  return props.active ? 'scale(1.2)' : 'scale(1)';
}, ";transition:all 0.2s cubic-bezier(0.175,0.885,0.32,1.275);");
exports.Handle = Handle;
