// @timberio/ui/dist/components/Ranger/Ranger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRanger = _interopRequireDefault(require("./react-ranger"));

var _components = require("./components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var defaultChildren = function defaultChildren(_ref) {
  var handleHeight = _ref.handleHeight,
      handleWidth = _ref.handleWidth,
      handleRadius = _ref.handleRadius,
      trackHeight = _ref.trackHeight,
      formatter = _ref.formatter;
  return function (_ref2) {
    var getTrackProps = _ref2.getTrackProps,
        ticks = _ref2.ticks,
        segments = _ref2.segments,
        handles = _ref2.handles;
    return _react.default.createElement(_components.Track, _extends({
      height: trackHeight
    }, getTrackProps({
      refProp: 'innerRef'
    })), ticks.map(function (_ref3) {
      var value = _ref3.value,
          getTickProps = _ref3.getTickProps;
      return _react.default.createElement(_components.Tick, _extends({
        key: "tick-".concat(value)
      }, getTickProps({
        refProp: 'innerRef'
      })), _react.default.createElement(_components.TickLabel, null, formatter(value)));
    }), segments.map(function (_ref4, i) {
      var value = _ref4.value,
          getSegmentProps = _ref4.getSegmentProps;
      return _react.default.createElement(_components.Segment, _extends({
        key: "segment-".concat(i)
      }, getSegmentProps({
        refProp: 'innerRef'
      }), {
        index: i
      }));
    }), handles.map(function (_ref5) {
      var value = _ref5.value,
          active = _ref5.active,
          getHandleProps = _ref5.getHandleProps;
      return _react.default.createElement("div", _extends({
        key: "handle-".concat(value)
      }, getHandleProps()), _react.default.createElement(_components.Handle, {
        active: active,
        height: handleHeight,
        width: handleWidth,
        radius: handleRadius
      }, formatter(value)));
    }));
  };
};

var Ranger =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Ranger, _React$Component);

  function Ranger() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Ranger);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Ranger)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      trackClicked: null,
      // eslint-disable-line
      trackClickedPrecise: null // eslint-disable-line

    });

    return _this;
  }

  _createClass(Ranger, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          value = _this$props.value,
          onChange = _this$props.onChange,
          stepSize = _this$props.stepSize,
          min = _this$props.min,
          max = _this$props.max,
          ticks = _this$props.ticks,
          steps = _this$props.steps,
          rest = _objectWithoutProperties(_this$props, ["value", "onChange", "stepSize", "min", "max", "ticks", "steps"]);

      return _react.default.createElement("div", null, _react.default.createElement(_reactRanger.default, {
        type: "linear",
        min: min,
        max: max,
        stepSize: stepSize,
        ticks: ticks,
        steps: steps,
        value: value,
        onChange: onChange
      }, defaultChildren(rest)));
    }
  }]);

  return Ranger;
}(_react.default.Component);

Ranger.defaultProps = {
  min: 0,
  max: 100,
  stepSize: 1,
  handleHeight: '30px',
  handleWidth: '30px',
  handleRadius: '100%',
  trackHeight: '8px',
  formatter: function formatter(v) {
    return v;
  }
};
Ranger.propTypes = {
  min: _propTypes.default.number,
  max: _propTypes.default.number,
  stepSize: _propTypes.default.number,
  handleHeight: _propTypes.default.string,
  handleWidth: _propTypes.default.string,
  handleRadius: _propTypes.default.string,
  trackHeight: _propTypes.default.string,
  formatter: _propTypes.default.func,
  ticks: _propTypes.default.array,
  steps: _propTypes.default.array,
  value: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.array])
};
var _default = Ranger;
exports.default = _default;
