// @timberio/ui/dist/helpers/animations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stroke = exports.scale = exports.fill = exports.flipUpY = exports.flipUpX = exports.slideIn = exports.rotate = exports.spin = exports.start = exports.progressLinearMovement = exports.fadeOutDown = exports.fadeInDown = exports.fadeOutUp = exports.fadeInUp = exports.fadeOut = exports.bounce = exports.bounceIn = exports.scaleIn = exports.fadeIn = void 0;

var _reactEmotion = require("react-emotion");

var fadeIn =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:0;}100%{opacity:1;}");
exports.fadeIn = fadeIn;
var scaleIn =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:0;transform:scale(.8) translateY(-30%);pointer-events:none;}100%{opacity:1;transform:none;pointer-events:auto;}");
exports.scaleIn = scaleIn;
var bounceIn =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{transform:scale(0.05);opacity:0;}55%{transform:scale(1.10);}75%{transform:scale(1.05);}100%{transform:scale(1);opacity:1;}");
exports.bounceIn = bounceIn;
var bounce =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{transform:scale(0.85);}55%{transform:scale(1.2);}75%{transform:scale(1.05);}100%{transform:scale(1);}");
exports.bounce = bounce;
var fadeOut =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:1;}100%{opacity:0;}");
exports.fadeOut = fadeOut;
var fadeInUp =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:0;position:relative;top:10px;}100%{opacity:1;position:relative;top:0;}");
exports.fadeInUp = fadeInUp;
var fadeOutUp =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:1;position:relative;top:0px;}100%{opacity:0;position:relative;top:-10px;}");
exports.fadeOutUp = fadeOutUp;
var fadeInDown =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:0;position:relative;bottom:10px;}100%{opacity:1;position:relative;bottom:0;}");
exports.fadeInDown = fadeInDown;
var fadeOutDown =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{opacity:1;position:relative;bottom:0;}100%{opacity:0;position:relative;bottom:-10px;}");
exports.fadeOutDown = fadeOutDown;
var progressLinearMovement =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{left:-100%;}50%{left:100%;}100%{left:100%;}");
exports.progressLinearMovement = progressLinearMovement;
var start =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("from{max-height:0;opacity:0;}to{max-height:20px;opacity:1;}");
exports.start = start;
var spin =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("to{transform:translateY(-15.0em);}");
exports.spin = spin;
var rotate =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("from{transform:rotate(0deg)}to{transform:rotate(360deg)}");
exports.rotate = rotate;
var slideIn =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{transform:translateY(-100%);opacity:0;}100%{transform:translateY(0%);opacity:1;}");
exports.slideIn = slideIn;
var flipUpX =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{transform:rotateX(-90deg) scale(0.9);}100%{transform:rotateX(0deg) scale(1);}");
exports.flipUpX = flipUpX;
var flipUpY =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%{transform:rotateY(-90deg) scale(0.9);}100%{transform:rotateY(0deg) scale(1);}");
exports.flipUpY = flipUpY;
var fill =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("100%{stroke-width:50;}");
exports.fill = fill;
var scale =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("0%,100%{transform:none;}50%{transform:scale3d(1.1,1.1,1);}");
exports.scale = scale;
var stroke =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("100%{stroke-dashoffset:0;}");
exports.stroke = stroke;
