// @timberio/ui/dist/helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "normalize", {
  enumerable: true,
  get: function get() {
    return _normalize2.default;
  }
});
Object.defineProperty(exports, "system", {
  enumerable: true,
  get: function get() {
    return _system2.default;
  }
});
Object.defineProperty(exports, "clean", {
  enumerable: true,
  get: function get() {
    return _clean2.default;
  }
});
Object.defineProperty(exports, "getRootEl", {
  enumerable: true,
  get: function get() {
    return _root.getRootEl;
  }
});
Object.defineProperty(exports, "INTENTS", {
  enumerable: true,
  get: function get() {
    return _theme.INTENTS;
  }
});
Object.defineProperty(exports, "SCALES", {
  enumerable: true,
  get: function get() {
    return _theme.SCALES;
  }
});
Object.defineProperty(exports, "propTypes", {
  enumerable: true,
  get: function get() {
    return _theme.propTypes;
  }
});
Object.defineProperty(exports, "fromTheme", {
  enumerable: true,
  get: function get() {
    return _theme.fromTheme;
  }
});
Object.defineProperty(exports, "intentColor", {
  enumerable: true,
  get: function get() {
    return _theme.intentColor;
  }
});
exports.animations = void 0;

var _normalize2 = _interopRequireDefault(require("./normalize"));

var _system2 = _interopRequireDefault(require("./system"));

var _clean2 = _interopRequireDefault(require("./clean"));

var _animations = _interopRequireWildcard(require("./animations"));

exports.animations = _animations;

var _root = require("./root");

var _theme = require("./theme");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
