// @timberio/ui/dist/components/Radio/Radio.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactEmotion = _interopRequireDefault(require("react-emotion"));

var _cleanElement = _interopRequireDefault(require("clean-element"));

var _objectPath = require("object-path");

var _helpers = require("../../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var cleaned = (0, _cleanElement.default)('div');
var Wrapper = (0, _helpers.system)(cleaned)(function (props) {
  return {
    display: 'inline-block',
    cursor: 'pointer'
  };
});
var Input =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "ez6pl160"
})(function (props) {
  return {
    display: 'inline-block',
    position: 'relative',
    width: '20px',
    height: '20px',
    verticalAlign: 'middle',
    boxSizing: 'border-box',
    borderRadius: '50%',
    border: "1px solid ".concat((0, _objectPath.get)(props, 'theme.colors.gray.4')),
    '&:after': {
      content: '""',
      position: 'absolute',
      top: '2px',
      bottom: '2px',
      left: '2px',
      right: '2px',
      borderRadius: '50%',
      background: props.checked ? (0, _objectPath.get)(props, 'theme.colors.purple.3') : 'none',
      transform: props.checked ? 'scale(1)' : 'scale(0)',
      animation: props.checked && props.animated ? "".concat(_helpers.animations.bounce, " 150ms ease-out") : 'none'
    }
  };
});
var Label = (0, _helpers.system)('span')(function (props) {
  return {
    verticalAlign: 'middle',
    marginLeft: '5px',
    color: (0, _objectPath.get)(props, 'theme.colors.gray.4')
  };
});

var Radio = function Radio(props) {
  var animated = props.animated,
      checked = props.checked,
      labelProps = props.labelProps,
      label = props.label,
      rest = _objectWithoutProperties(props, ["animated", "checked", "labelProps", "label"]);

  return _react.default.createElement(Wrapper, rest, _react.default.createElement(Input, {
    animated: animated,
    checked: checked
  }), _react.default.createElement(Label, labelProps, label));
};

Radio.defaultProps = {
  animated: true,
  checked: false,
  labelProps: {}
};
Radio.propTypes = {
  animated: _propTypes.default.bool,
  checked: _propTypes.default.bool,
  labelProps: _propTypes.default.object,
  label: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
};
cleaned.propTypes = Radio.propTypes;
Radio.displayName = 'Radio';
var _default = Radio;
exports.default = _default;
