// @timberio/ui/dist/components/Indicator/Indicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testExports = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactEmotion = _interopRequireWildcard(require("react-emotion"));

var _helpers = require("../../helpers");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var scaleFadeOut =
/*#__PURE__*/
(0, _reactEmotion.keyframes)("from{transform:scale(0);opacity:0.5;}to{transform:scale(3);opacity:0;}");
var propTypes = {
  intent: _helpers.propTypes.intents
};
var IndicatorContainer =
/*#__PURE__*/
(0, _reactEmotion.default)((0, _helpers.system)('div', propTypes)(), {
  target: "e1vgypy20"
})("display:inline-block;position:relative;width:", function (_ref) {
  var size = _ref.size;
  return size;
}, "px;height:", function (_ref2) {
  var size = _ref2.size;
  return size;
}, "px;border-radius:50%;background-color:", function (props) {
  return (0, _helpers.intentColor)(props, {
    shade: 2
  });
}, ";");
var AnimatedDot =
/*#__PURE__*/
(0, _reactEmotion.default)('div', {
  target: "e1vgypy21"
})("position:absolute;width:100%;height:100%;border-radius:50%;background-color:", function (props) {
  return (0, _helpers.intentColor)(props, {
    shade: 2
  });
}, ";animation:", scaleFadeOut, " 1.5s ease-in-out infinite;animation-delay:", function (_ref3) {
  var delay = _ref3.delay;
  return delay || 0;
}, "ms;");

var Indicator = function Indicator(_ref4) {
  var intent = _ref4.intent,
      rest = _objectWithoutProperties(_ref4, ["intent"]);

  return _react.default.createElement(IndicatorContainer, _extends({}, rest, {
    intent: intent
  }), _react.default.createElement(AnimatedDot, {
    intent: intent
  }), _react.default.createElement(AnimatedDot, {
    intent: intent,
    delay: 300
  }));
};

Indicator.propTypes = propTypes;
Indicator.defaultProps = {
  intent: 'primary',
  size: 12
};
var _default = Indicator;
exports.default = _default;
var testExports = {
  IndicatorContainer: IndicatorContainer,
  AnimatedDot: AnimatedDot
};
exports.testExports = testExports;
