// @timberio/ui/dist/components/Spinner/Spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _cleanElement = _interopRequireDefault(require("clean-element"));

var _helpers = require("../../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function speedSwitch(speed) {
  if (speed === 'fast') return 600;
  if (speed === 'slow') return 900;
  return 750;
}

var cleaned = (0, _cleanElement.default)('div');
var Wrapper = (0, _helpers.system)(cleaned)(function (props) {
  return {
    animationName: _helpers.animations.rotate,
    animationIterationCount: 'infinite',
    animationTimingFunction: 'linear',
    animationDuration: "".concat(speedSwitch(props.speed), "ms"),
    height: props.size,
    width: props.size,
    color: props.color ? props.color : (0, _helpers.intentColor)(props, {
      intent: props.intent,
      shade: 3
    })
  };
});

var Spinner = function Spinner(props) {
  var gap = props.gap,
      thickness = props.thickness,
      rest = _objectWithoutProperties(props, ["gap", "thickness"]);

  return _react.default.createElement(Wrapper, rest, _react.default.createElement("svg", {
    role: "img",
    "aria-labelledby": "title desc",
    viewBox: "0 0 32 32",
    stroke: "currentColor"
  }, _react.default.createElement("circle", {
    role: "presentation",
    cx: 16,
    cy: 16,
    r: 14 - thickness / 2,
    fill: "none",
    strokeWidth: thickness,
    strokeDasharray: Math.PI * 2 * (11 - gap),
    strokeLinecap: "round"
  })));
};

Spinner.defaultProps = {
  intent: 'primary',
  gap: 4,
  thickness: 4,
  size: 25,
  speed: 'fast'
};
Spinner.propTypes = {
  intent: _helpers.propTypes.intents,
  thickness: _propTypes.default.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  gap: _propTypes.default.oneOf([1, 2, 3, 4, 5]),
  speed: _propTypes.default.oneOf(['fast', 'slow'])
};
cleaned.propTypes = Spinner.propTypes;
Spinner.displayName = 'Spinner';
var _default = Spinner;
exports.default = _default;
