// @timberio/ui/dist/components/Image/Image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _emotionTheming = require("emotion-theming");

var _cleanElement = _interopRequireDefault(require("clean-element"));

var _md = _interopRequireDefault(require("md5"));

var _unfetch = _interopRequireDefault(require("unfetch"));

var _objectPath = require("object-path");

var _helpers = require("../../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var cleaned = (0, _cleanElement.default)('img');
var Wrapper = (0, _helpers.system)(cleaned)(function (props) {
  return {
    display: 'inline-block',
    background: (0, _objectPath.get)(props, 'theme.colors.gray.2'),
    color: (0, _objectPath.get)(props, 'theme.colors.gray.5')
  };
});

var Image =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Image, _PureComponent);

  function Image(_props) {
    var _this;

    _classCallCheck(this, Image);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Image).call(this, _props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "parseProvider", function () {
      var provider = _this.props.provider;

      if (provider === 'placeholder') {
        _this.getPlaceholder();
      }

      if (provider === 'picsum') {
        _this.getPicsumUrl();
      }

      if (provider === 'kitten') {
        _this.getPlacekittenUrl();
      }

      if (provider === 'freeman') {
        _this.getFreemanUrl();
      }

      if (provider === 'gravatar') {
        _this.getGravatarUrl();
      }

      if (provider === 'clearbit') {
        _this.getClearbitUrl();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "normalizeSize", function (retina) {
      var _this$props = _this.props,
          width = _this$props.width,
          height = _this$props.height,
          size = _this$props.size,
          props = _objectWithoutProperties(_this$props, ["width", "height", "size"]);

      if (size) {
        var val = Array.isArray(size) ? size[size.length - 1] : size;
        return retina ? [val * 2, val * 2, props] : [val, val, props];
      }

      return retina ? [width * 2, height * 2, props] : [width, height, props];
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onError", function () {
      _this.getPlaceholder(_this.props.query.charAt(0).toUpperCase());
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getPlaceholder", function (value) {
      var _this$props2 = _this.props,
          theme = _this$props2.theme,
          _this$props2$query = _this$props2.query,
          query = _this$props2$query === void 0 ? ' ' : _this$props2$query,
          placeholder = _this$props2.placeholder;

      var _this$normalizeSize = _this.normalizeSize(true),
          _this$normalizeSize2 = _slicedToArray(_this$normalizeSize, 2),
          width = _this$normalizeSize2[0],
          height = _this$normalizeSize2[1]; // If a placeholder component was passed, use that instead of an image


      if (placeholder) {
        return _this.setState({
          placeholder: true
        });
      } // Otherwise generate a placeholder image


      var text = value || query.replace(/\s/g, '+');
      var color = (0, _objectPath.get)(theme, 'colors.gray.2').replace(/^#/, '');

      _this.setState({
        url: "https://placehold.it/".concat(width, "x").concat(height, "/").concat(color, "/?text=").concat(text)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getPicsumUrl", function () {
      var _this$normalizeSize3 = _this.normalizeSize(true),
          _this$normalizeSize4 = _slicedToArray(_this$normalizeSize3, 2),
          width = _this$normalizeSize4[0],
          height = _this$normalizeSize4[1];

      _this.setState({
        url: "https://picsum.photos/".concat(width, "/").concat(height)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getPlacekittenUrl", function () {
      var _this$normalizeSize5 = _this.normalizeSize(true),
          _this$normalizeSize6 = _slicedToArray(_this$normalizeSize5, 2),
          width = _this$normalizeSize6[0],
          height = _this$normalizeSize6[1];

      _this.setState({
        url: "https://placekitten.com/".concat(width, "/").concat(height)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getFreemanUrl", function () {
      var _this$normalizeSize7 = _this.normalizeSize(true),
          _this$normalizeSize8 = _slicedToArray(_this$normalizeSize7, 2),
          width = _this$normalizeSize8[0],
          height = _this$normalizeSize8[1];

      _this.setState({
        url: "https://morganfillman.space/".concat(width, "/").concat(height)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getGravatarUrl", function () {
      var query = _this.props.query;

      var _this$normalizeSize9 = _this.normalizeSize(true),
          _this$normalizeSize10 = _slicedToArray(_this$normalizeSize9, 1),
          width = _this$normalizeSize10[0];

      _this.setState({
        url: "https://www.gravatar.com/avatar/".concat((0, _md.default)(query), "?s=").concat(width)
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getClearbitUrl", function () {
      var query = _this.props.query;
      (0, _unfetch.default)("https://autocomplete.clearbit.com/v1/companies/suggest?query=".concat(query)).then(function (res) {
        return res.ok ? res.json() : Promise.reject(res);
      }).then(function (data) {
        if (Array.isArray(data) && _typeof(data[0]) === 'object') {
          _this.setState({
            url: data[0].logo
          });
        } else {
          _this.onError();
        }
      }).catch(function (err) {
        _this.onError();

        console.error('Could not fetch Clearbit logo!', err);
      });
    });

    _this.state = {
      url: _props.url,
      placeholder: false
    };
    return _this;
  }

  _createClass(Image, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.parseProvider();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.query !== prevProps.query) {
        this.parseProvider();
      }
    }
  }, {
    key: "renderPlaceholder",
    value: function renderPlaceholder() {
      var _this$props3 = this.props,
          placeholder = _this$props3.placeholder,
          query = _this$props3.query;

      var _this$normalizeSize11 = this.normalizeSize(true),
          _this$normalizeSize12 = _slicedToArray(_this$normalizeSize11, 2),
          width = _this$normalizeSize12[0],
          height = _this$normalizeSize12[1];

      return _react.default.createElement(placeholder, {
        width: width,
        height: height,
        query: query
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$normalizeSize13 = this.normalizeSize(),
          _this$normalizeSize14 = _slicedToArray(_this$normalizeSize13, 3),
          width = _this$normalizeSize14[0],
          height = _this$normalizeSize14[1],
          props = _this$normalizeSize14[2];

      if (this.state.placeholder) {
        return this.renderPlaceholder();
      }

      return _react.default.createElement(Wrapper, _extends({}, props, {
        width: width,
        height: height,
        src: this.state.url,
        onError: this.onError
      }));
    }
  }]);

  return Image;
}(_react.PureComponent);

Image.defaultProps = {
  url: null,
  query: null,
  provider: 'picsum',
  // styled-system
  width: 100,
  height: 100
};
Image.propTypes = {
  url: _propTypes.default.string,
  query: _propTypes.default.string,
  placeholder: _propTypes.default.func,
  provider: _propTypes.default.oneOf(['placeholder', 'picsum', 'kitten', 'freeman', 'gravatar', 'clearbit'])
};
cleaned.propTypes = Image.propTypes;
Image.displayName = 'Image';

var _default = (0, _emotionTheming.withTheme)(Image);

exports.default = _default;
