// @timberio/ui/dist/components/Show/Show.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactShow = _interopRequireDefault(require("react-show"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var transitions = {
  height: {
    hide: {
      opacity: 0,
      height: 0
    },
    show: {
      opacity: 1,
      height: 'auto'
    }
  },
  width: {
    hide: {
      opacity: 0,
      width: 0
    },
    show: {
      opacity: 1,
      width: 'auto'
    }
  }
};

var Show = function Show(_ref) {
  var show = _ref.show,
      duration = _ref.duration,
      type = _ref.type,
      children = _ref.children;
  return _react.default.createElement(_reactShow.default, {
    show: show,
    duration: duration,
    styleHide: transitions[type].hide,
    styleShow: transitions[type].show
  }, children);
};

Show.propTypes = {
  show: _propTypes.default.bool.isRequired,
  children: _propTypes.default.node.isRequired,
  type: _propTypes.default.oneOf(['height', 'width']),
  duration: _propTypes.default.number
};
Show.defaultProps = {
  type: 'height',
  duration: 200
};
var _default = Show;
exports.default = _default;
