// @timberio/ui/dist/components/Callout/Callout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testExports = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactEmotion = _interopRequireDefault(require("react-emotion"));

var _objectPath = require("object-path");

var _Icon = _interopRequireDefault(require("../Icon"));

var _helpers = require("../../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var propTypes = {
  intent: _helpers.propTypes.intents,
  icon: _propTypes.default.string
};
var Container =
/*#__PURE__*/
(0, _reactEmotion.default)((0, _helpers.system)('div', propTypes)(), {
  target: "epm9cpm0"
})("width:100%;display:flex;align-items:center;font-size:", function (props) {
  return (0, _objectPath.get)(props, 'theme.fontSizes.3');
}, ";background-color:", function (props) {
  return (0, _helpers.intentColor)(props, {
    shade: 0
  });
}, ";color:", function (props) {
  return (0, _helpers.intentColor)(props, {
    shade: 5
  });
}, ";border:1px solid ", function (props) {
  return (0, _helpers.intentColor)(props, {
    shade: 3
  });
}, ";border-radius:4px;");

var Callout = function Callout(_ref) {
  var intent = _ref.intent,
      children = _ref.children,
      icon = _ref.icon,
      onClick = _ref.onClick,
      rest = _objectWithoutProperties(_ref, ["intent", "children", "icon", "onClick"]);

  return _react.default.createElement(Container, _extends({
    intent: intent
  }, rest), icon && _react.default.createElement(_Icon.default, {
    mr: 10,
    name: icon,
    size: 20
  }), children);
};

Callout.defaultProps = {
  intent: 'primary',
  buttonText: 'Confirm',
  p: 3
};
Callout.displayName = 'Callout';
var _default = Callout;
exports.default = _default;
var testExports = {
  Container: Container
};
exports.testExports = testExports;
